### PR TITLE
docs: Integration section and tutorial

### PR DESCRIPTION
**Pull request by bpmutter**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

- Add Integrate with ESLint section to docs
- create an integrations with Node.js API tutorial

#### Is there anything you'd like reviewers to focus on?

Resolves https://github.com/eslint/eslint/issues/17134
Resolves https://github.com/eslint/eslint/issues/17133

### CODE REVIEW

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [inline docs/src/integrate/integration-tutorial.md:1]: @nzakas before putting this tutorial up for formal review, do you like the approach taken here? any higher-level feedback you'd like implemented?

**Comment by ollie-iterators** [comment]
Errors with verification of files:

- docs/src/integrate/integration-tutorial.md: 16: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 1[8](https://github.com/eslint/eslint/actions/runs/4867743708/jobs/8680605325#step:5:9): MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 20: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 22: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 24: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 36: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 37: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 38: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
- docs/src/integrate/integration-tutorial.md: 6[9](https://github.com/eslint/eslint/actions/runs/4867743708/jobs/8680605325#step:5:10): MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
- docs/src/integrate/integration-tutorial.md: [13](https://github.com/eslint/eslint/actions/runs/4867743708/jobs/8680605325#step:5:14)7: MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
- docs/src/integrate/integration-tutorial.md: 252: MD047/single-trailing-newline Files should end with a single newline character

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (nzakas) [inline docs/_integration-tutorial-code/example-eslint-integration.js:17]: `getLintingResults` doesn't seem like the right name here because we are doing more than just retrieve a value. (Running ESLint and also writing fixes to disk.)

Maybe just `lintAndFix()`?

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (nzakas) [inline docs/_integration-tutorial-code/example-eslint-integration.js:16]:
```suggestion
// a function that lints the specified files and returns the error results
```

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (nzakas) [inline docs/_integration-tutorial-code/example-eslint-integration.js:24]: Is there a reason you're filtering out warnings here?

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (nzakas) [inline docs/_integration-tutorial-code/example-eslint-integration.js:30]: Maybe a different name would make this clearer?
```suggestion
function outputLintingResults(results) {
```

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (nzakas) [inline docs/_integration-tutorial-code/example-eslint-integration.js:40]:
```suggestion
// Put previous functions all together
```

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (nzakas) [inline docs/_integration-tutorial-code/example-eslint-integration.js:8]: I don't think this function is necessary because you're only ever using one branch of it.

File: docs/src/integrate/index.md
Comment (nzakas) [inline docs/src/integrate/index.md:10]:
```suggestion
This guide is intended for those who wish to integrate the functionality of ESLint into other applications by using the ESLint API.
```

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:12]: We don't use the word "custom" anywhere else.

```suggestion
## Why Create an Integration?
```

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:18]: I'd need to double-check, but I think most CI systems use the CLI.

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:25]: I'd also add plugins for code-related tools like Webpack and Jest.

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:93]: This text feels a bit disjointed to me and disconnected from the following code example. Can you expand this to talk a bit more about what each method does? For instance, lintFiles searches the filesystem and the file paths can be globs or filenames.

Also, outputFixes isn't for readability, it's the method that outputs fixes back to disk. getErrorResults filters out all warnings.

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:116]: Again, "custom" isn't necessary. 
```suggestion
Define a function to output the linting results to the consoles. This should be specific to your integration's needs.
```

**Review by nzakas — COMMENTED** [review]
I think the overall approach is helpful, but some of the naming and details are incorrect. We can keep iterating on those details.

Overall, I think this is a good way to introduce how to use the Node.js API in a way that will be useful to most folks.

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (bpmutter) [reply docs/_integration-tutorial-code/example-eslint-integration.js:24]: no, will look at how to capture warnings as well, and if cannot figure out will ask for further assistance

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md:18]: ok, i can remove this bullet point, then

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md:25]: adding a line for 'developer tools'

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (bpmutter) [reply docs/_integration-tutorial-code/example-eslint-integration.js:17]: changing

File: docs/_integration-tutorial-code/example-eslint-integration.js
Comment (bpmutter) [reply docs/_integration-tutorial-code/example-eslint-integration.js:8]: keeping the fxn b/c i think it's useful for the tutorial flow, but simplifying to not have the conditional, and just return the instance:

```js
// Create an instance of ESLint with the configuration passed to the function
function createESLintInstance(overrideConfig){
  return new ESLint({ useEslintrc: false, overrideConfig: overrideConfig });
}
```

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md]: updated copy here

**Comment by Rec0iL99** [comment]
Not stale. Waiting for @nzakas's review.

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [reply docs/src/integrate/integration-tutorial.md:25]: I think we're still missing the Webpack/Jest plugin use case. In that case you're not creating a new tool but rather extending an existing tool. I think we need to be more explicit about this use case.

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:37]: Yarn isn't used anywhere on the page, so I think we can safely omit it.
```suggestion
* npm
```

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:55]: It may be worth calling out that the package is installed as a dependency and not a dev dependency for this use case.

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:84]: We're not just retrieving results here, so maybe this would be a better title?

```suggestion
## Step 3: Lint and Fix Files
```

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:88]:
```suggestion
The static method `ESLint.outputFixes()` takes the linting results from the call to `ESLint#lintFiles()`, and then writes the fixed code back to the source files.
```

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:106]:
```suggestion
## Step 4: Output Results
```

File: docs/src/integrate/integration-tutorial.md
Comment (nzakas) [inline docs/src/integrate/integration-tutorial.md:230]: Rather than creating a top-level directory under `docs`, maybe we should have `docs/_examples` for all examples and then a subdirectory under `docs/_examples` for the integration example code?

**Review by nzakas — CHANGES_REQUESTED** [review]
This is looking good! Just a few suggestions throughout.

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md:25]: updated the bullet point. lmk what you think of the update

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md]: updated

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md]: moved

**Comment by bpmutter** [comment]
> [CLA Missing ID](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/19114526/11061773/17132/#/?version=2) [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/19114526/11061773/17132/#/?version=2)
> 
> * ✅ login: bpmutter / name: Ben Perlmutter  ([8b4b4a5](https://github.com/eslint/eslint/commit/8b4b4a5e3085e7e850294ee0c3cb059f274ca554), [2572ea4](https://github.com/eslint/eslint/commit/2572ea40f29ae7f35265d9eda06b08af3e9db376), [4fadcfb](https://github.com/eslint/eslint/commit/4fadcfbe227c449e6a19858f552ecf6affb2b4ef), [e8953f6](https://github.com/eslint/eslint/commit/e8953f68e5b2487a783eaa4b176bd41047868072), [5123dfd](https://github.com/eslint/eslint/commit/5123dfd1b509e58074417114238e8b07180ac210), [d8770ae](https://github.com/eslint/eslint/commit/d8770ae7d7745d6c13717ca5ee3a78f699508cef), [487bac6](https://github.com/eslint/eslint/commit/487bac6339645dfcb52a6c5f1ce19591290479d9), [8032044](https://github.com/eslint/eslint/commit/803204462b6e703f0e2c330be17631fb202dd2a9), [88a1dcb](https://github.com/eslint/eslint/commit/88a1dcb6f1f02468e81528d78279986cca805bd1), [d40cc3a](https://github.com/eslint/eslint/commit/d40cc3a01fadb32386369a86c3776a4a444ff2f2), [941d689](https://github.com/eslint/eslint/commit/941d689655186f421fc35a45be29b73e870b5900), [9686b50](https://github.com/eslint/eslint/commit/9686b50fc3192cff61ea1428f799b7e8b4eacd3d), [aaefddb](https://github.com/eslint/eslint/commit/aaefddb05cec78761e6f4b4447cc602f4a363696))
> * ❌ The email address for the commit ([487bac6](https://github.com/eslint/eslint/commit/487bac6339645dfcb52a6c5f1ce19591290479d9), [9686b50](https://github.com/eslint/eslint/commit/9686b50fc3192cff61ea1428f799b7e8b4eacd3d)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@nzakas i think this might be something you have to do on your end. on these 2 commits which the bot has flagged, you're coauthor b/c i accepted your suggestion

**Review by nzakas — APPROVED** [review]
LGTM. Can you rebase on top of main to pull in our updated docs license?

**Comment by nzakas** [comment]
I think the CLA bot might have a bug. The email address on my suggestion is the same one I always use and it's definitely tied to my GitHub account. I've opened a bug with EasyCLA.

**Comment by bpmutter** [comment]
rebased!

**Review by snitin315 — COMMENTED** [review]
Looks good overall. A few suggestions.

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md:64]: no, it wouldn't (well not unless you're using linux subsystem). 

personally i'm comfortable w this as is given the popularity/knowledge of unix-like systems and context of the above text that talks about creating a file. 

but fine adjusting to something like:

```shell
# UNIX-like systems
touch example-eslint-integration.js

# Windows
type NUL > example-eslint-integration.js
```

i feel this is a little clunky, though. and think it's clearer to just keep as is

Comment (nzakas) [reply docs/src/integrate/integration-tutorial.md:64]: I think `touch` fine, it's fairly well known.

**Comment by nzakas** [comment]
@mdjermanovic do you want to review this before merging?

**Comment by mdjermanovic** [comment]
> @mdjermanovic do you want to review this before merging?

I'll take a look now.

File: docs/_examples/integration-tutorial-code/sample-data/test-file.js
Comment (mdjermanovic) [inline docs/_examples/integration-tutorial-code/sample-data/test-file.js:20]: Same as above, `indent` is not in `eslint:recommended` and it isn't enabled in the hardcoded configuration inside `example-eslint-integration`.

File: docs/_examples/integration-tutorial-code/package.json
Comment (mdjermanovic) [inline docs/_examples/integration-tutorial-code/package.json:15]: Do we need these in the example project?

File: docs/_examples/integration-tutorial-code/.eslintrc.js
Comment (mdjermanovic) [inline docs/_examples/integration-tutorial-code/.eslintrc.js:1]: I think this file isn't used in any way, so we could remove it?

File: docs/_examples/integration-tutorial-code/sample-data/test-file.js
Comment (bpmutter) [inline docs/_examples/integration-tutorial-code/sample-data/test-file.js:7]:
```suggestion
```

File: docs/_examples/integration-tutorial-code/sample-data/test-file.js
Comment (bpmutter) [reply docs/_examples/integration-tutorial-code/sample-data/test-file.js:20]: same comment as above. don't remember context. deleting the example

File: docs/src/integrate/integration-tutorial.md
Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md:188]: in this case, could you please provide an example of what would be correct here?

Comment (bpmutter) [reply docs/src/integrate/integration-tutorial.md:188]: adjusting. pls lmk what you think of the new approach

File: docs/_examples/integration-tutorial-code/package.json
Comment (bpmutter) [reply docs/_examples/integration-tutorial-code/package.json:15]: removing. 

i had these and the .eslintrc file b/c i was using it to lint the code i was writing

**Comment by bpmutter** [comment]
@mdjermanovic ready for re-review

@nzakas accidentally re-requested review. sorry about that. no action necessary right now

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
